### PR TITLE
Support function overloading

### PR DESCRIPTION
**Pull request by YannanGao-gs**
**What type of PR is this?**
Feature Request

**What does this PR do / why is it needed ?**
update the path of Function from package::name to package::function signature

1. Update Engine parser so that the function signature will be used for function.name
2. SDLC identifies the old style (use functionName as the file name) functions, then creates a new entity that has functionName as its path  
3. Studio will create a ‘Delete’ action to delete this kind of entity above, then do a ‘Create’ operation to correct the gitlab file name
 

**Which issue(s) this PR fixes:**
Fixes #

**Other notes for reviewers:
Does this PR introduce a user-facing change?**
Related to https://github.com/finos/legend-sdlc/pull/515 and https://github.com/finos/legend-studio/pull/1339
Require Alloy changes

### CODE REVIEW

File: legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java [hunk 1]
```diff
@@ -596,7 +596,7 @@ public void testMetaNewFunctionWithSingleParameter()
     @Test
     public void testMultiIFExpressions()
     {
-        test("function test::multiExpressions(): String[0..1]\n" +
```
Comment (MauricioUyaguari) [inline legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java:599]: 💀 ☠️

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

**Comment by YannanGao-gs** [comment]
Need to fix broken tests

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java [hunk 2]
```diff
@@ -395,6 +397,10 @@ private org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain
         func.returnType = ctx.functionTypeSignature().type().getText();
         func.returnMultiplicity = this.buildMultiplicity(ctx.functionTypeSignature().multiplicity().multiplicityArgument());
         func.sourceInformation = this.walkerSourceInformation.getSourceInformation(ctx);
+        String functionSignature =   LazyIterate.collect(func.parameters, HelperValueSpecificationGrammarComposer::getParameterSignature).select(Objects::nonNull).makeString("__")
```
Comment (MauricioUyaguari) [inline legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java:400]: Is this change needed because when going from text to pmcd we can't have 2 elements with the same package and name?
Do you know where that constraint comes from?
My thinking is that the grammar/protocol should not have to handle this. 
It is when we compile (protocol -> metamodel)  the function we need to provide and generate the full function name.

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/HelperModelBuilder.java [hunk 3]
```diff
@@ -279,10 +271,10 @@ public static String getTerseSignature(Function function)
 
     private static String terseSignatureSuffix(Function function)
     {
-        return "_" + LazyIterate.collect(function.parameters, HelperModelBuilder::getParameterSignature).select(Objects::nonNull).makeString("__")
+        String functionSignature =  LazyIterate.collect(function.parameters, HelperModelBuilder::getParameterSignature).select(Objects::nonNull).makeString("__")
```
Comment (MauricioUyaguari) [inline legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/HelperModelBuilder.java:274]: Could we make the logic more clear here. Maybe use string builders. Also lets add tests for these functions to make it clear what the expectation that each returns.

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java [hunk 4]
```diff
@@ -395,6 +397,10 @@ private org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain
...
+            + "__" + HelperValueSpecificationGrammarComposer.getClassSignature(func.returnType) + "_" + HelperValueSpecificationGrammarComposer.getMultiplicitySignature(func.returnMultiplicity) + "_";
+        String terseSignatureSuffix =  func.parameters.size() > 0 ? "_" + functionSignature : functionSignature;
+        func.name = PureGrammarParserUtility.fromIdentifier(ctx.qualifiedName().identifier()) + terseSignatureSuffix;
```
Comment (MauricioUyaguari) [inline legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java:403]: lets move this logic out and call it like `buildFullFunctionSignature` to make it clear we are assigning the full function name to the .name property.

Comment (MauricioUyaguari) [reply legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java:403]: Dont we also have already defined methods we could use here like `getFunctionSignature` and getFunctionName ?

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/HelperModelBuilder.java (hunk 3 above)
Comment (YannanGao-gs) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/HelperModelBuilder.java:274]: tests are added

**Comment by akphi** [comment]
@YannanGao-gs committing .DS_Store file (system files) is a no-no. `engine` `gitignore` should prevent you from doing this, but it didn't. But you should remove that from the commit

**Review by akphi — APPROVED** [review]
Looks good to me

**Review by akphi — CHANGES_REQUESTED** [review]
Except for the problem with `DS_Store` file mentioned.

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/to/HelperValueSpecificationGrammarComposer.java [hunk 5]
```diff
@@ -250,23 +251,26 @@ public static String printFullPath(String fullPath, DEPRECATED_PureGrammarCompos
     public static String getFunctionName(org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain.Function fn)
     {
         int signatureIndex = fn.name.indexOf(getFunctionSignature(fn));
-        String name = signatureIndex > 0 ? fn.name.substring(0, signatureIndex) : fn.name;
-        return fn._package == null || fn._package.isEmpty() ? name : fn._package + "::" + name;
+        if (signatureIndex <= 0)
```
Comment (MauricioUyaguari) [inline legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/to/HelperValueSpecificationGrammarComposer.java:254]: let's remove this throw for now.

**Review by pierredebelen — APPROVED** [review]
